src/bin/pg_autoctl/cli_create_drop_node.c

```diff
@@ -753,6 +753,25 @@ cli_create_monitor(int argc, char **argv)
 
 	monitor.config = monitorOptions;
 
+
+	/*
+	 * We prefer to ignore PGHOST variable for the monitor. The reason is that we
+	 * qualify monitor database as an internal/embedded database for pg_auto_failover.
+	 * In other words, we don't want anyone else to be able to connect to the database,
+	 * except for pg_auto_failover itself. To do that, we hard code host to be the
+	 * unix domain socket, where we can restrict the access.
+	 *
+	 * TODO: Not allowing to set PGHOST fails regression tests with:
+	 *
+	 * 	ERROR pgsetup.c:269 PG_REGRESS_SOCK_DIR is set to "" to disable
+	 * 	unix socket directories, now --pghost is mandatory, but unset.
+	 *
+	 */
+	if (!env_found_empty("PG_REGRESS_SOCK_DIR")  && unsetenv("PGHOST") != 0)
+	{
+		exit(EXIT_CODE_BAD_STATE);
+	}
+
```

Comment:
I think this comment and code does not reflect the decision we made entirely. On top of that, the decision we made is not compatible with the current coding of the test-suite, and that needs to be taken care of in the test suite itself, as you started to discover.

A better comment here might be:

``` c
	/*
	 * The pg_auto_failover monitor Postgres database needs to allow
	 * connections from the pg_autoctl monitor binary itself (this code) and
	 * later from other pg_autoctl nodes that are using this monitor instance.
	 *
	 * In any case, we don't use --pghost nor PGHOST to build our connection
	 * string to the Postgres monitor database in this `pg_autoctl create
	 * monitor` program, instead we always use the Unix Domain Socket directory
	 * used by Postgres.
	 */
```

Then in our testing environment we set PG_REGRESS_SOCK_DIR to an empty string. In that we imitate what `pg_regress` does at https://github.com/postgres/postgres/blob/master/src/test/regress/pg_regress.c#L802, for much the same reasons: allow to start several Postgres instances on the same port on the same host. In our test suite, we use network namespaces so that each Postgres instance has their own IP address to listen to, and a single file system, so that we can't use unix domain sockets at the moment: we would try to create the same files for all the nodes (same default Postgres directory, same port).

I think we should mimic what `pg_regress` is doing in `make_temp_sockdir` at https://github.com/postgres/postgres/blob/master/src/test/regress/pg_regress.c#L318 when PG_REGRESS_SOCK_DIR is empty and use the temporary directory created both in the `postmaster -k` option and then in our connection strings, so in our `pgSetup.pghost` structure too. We could implement that part by forcing the `PGHOST` value to the temp sockdir we just created and then `pgsetup.c` will take it from there.

Given that, we should be able to connect to the monitor Postgres database from the test suite without exposing a unix socket directory at the default place, which would then be re-used by every other Postgres instance, and because they all use the same port in our test-suite, that would fail.

Reply:
There are few things that I cannot easily follow here which makes it hard for me to implement

> Then in our testing environment we set PG_REGRESS_SOCK_DIR to an empty string. In that we imitate what pg_regress does at https://github.com/postgres/postgres/blob/master/src/test/regress/pg_regress.c#L802, for much the same reasons:


When I read the Postgres  code/commit that introduces the env. variable, the main motivation seems different: https://github.com/postgres/postgres/commit/31c6e54ec9abab0c63d709e492ef18a701b02641

>  allow to start several Postgres instances on the same port on the same host.

Is that possible with this? I'm trying to do that locally, but I cannot see a way for doing _that other than setting `listen_addresses ` to empty list.


> We could implement that part by forcing the PGHOST value to the temp sockdir we just created and then pgsetup.c will take it from there.

Could you please clarify this a little further? I though we're trying to unset `PGHOST`, how can we force it?

Reply:
1. ah right it was security for `pg_regress`, for the build farm animals I suppose
  2. that's possible if you have ethernet interface aliases or network namespace, the latter is what we do in our test suite ; then each Postgres instance can listen to its own ethernet interface (which just happens even with `listen_addresses = '*'` with network namespaces)
  3. we never said we wanted to unset `PGHOST`, only that we don't care about the value that the user has set in their environment

Reply:
> we never said we wanted to unset PGHOST, only that we don't care about the value that the user has set in their environment

Hmm, my fear is that if we don't unset it, we might end-up with a state where in some  parts of the code we rely on `PGHOST`, in some it doesn't. That way, we might lead to a discrepancy which would be hard to understand by  anyone.  Sort of creating a hard to track dependency. 

That's why I considered this as we should "unset" it. One  of the reasons for me to go with "unset" approach is that I don't have a good enough visibility of the code-base at this moment. But, if you think it is safe just ignore here and let other parts rely on it (if any), I can  do that.

Reply:
Thanks to this discussion what I now have in mind is that the `pg_autoctl create monitor` command, when PG_REGRESS_SOCK_DIR is set empty in the environment, would then create a temporary socket directory and use it in the `-k` option used when starting Postgres in either `pg_ctl_start` or `pg_ctl_postgres`.

Once we've done that though, we still need to generate a local connection string that uses that temp socket directory as our `host` option, as in `psql -h /tmp -p 5001 -d postgres` if you have a local server running on port 5001 with unix sockets created in the `/tmp` directory.

The easiest way to achieve that given our current code-base (see `pgsetup.c`) might be to force set the PGHOST environment variable to that temporary directory internally, and make sure to call `pg_setup_init` again after that.